libqtile/window.py

```diff
@@ -727,17 +727,45 @@ def update_strut(self):
             "_NET_WM_STRUT_PARTIAL",
             unpack=int
         )
-        strut = strut or self.window.get_property(
-            "_NET_WM_STRUT",
-            unpack=int
-        )
         if strut:
-            self.qtile.add_strut(strut)
-        self.strut = strut
+            x_screen_dimensions = self.qtile.root.get_geometry()
+            if strut[0]:    # left
+                x = strut[0]
+                y = (strut[4] + strut[5]) / 2 or (strut[6] + strut[7]) / 2
+            elif strut[1]:  # right
+                x = x_screen_dimensions.width - strut[1]
+                y = (strut[4] + strut[5]) / 2 or (strut[6] + strut[7]) / 2
+            elif strut[2]:  # top
+                x = (strut[8] + strut[9]) / 2 or (strut[10] + strut[11]) / 2
+                y = strut[2]
+            else:           # bottom
+                x = (strut[8] + strut[9]) / 2 or (strut[10] + strut[11]) / 2
+                y = x_screen_dimensions.height - strut[3]
+            self.screen = self.qtile.find_screen(x, y)
+
+            if self.screen is None:
+                logger.error("No screen at target")
+                return
+            elif None in [self.screen.x, self.screen.y, self.screen.height, self.screen.width]:
+                logger.error("Missing screen information")
+                return
+
+            empty_space = [
+                self.screen.x,
+                x_screen_dimensions.width - self.screen.x - self.screen.width,
+                self.screen.y,
+                x_screen_dimensions.height - self.screen.y - self.screen.height
+            ]
+            self.reserved_space = [strut[i] - empty if strut[i] else 0 for i, empty in enumerate(empty_space)]
```

Comment:
I'm finding it a bit hard to see what's going on here. What are the values in `empty_space`?

The rest of the PR look great, thanks

Reply:
`empty_space` is what used to be called `gaps` in earlier commits, I changed the name as the similarity with `bar.Gap` started to bug me. It contains dimensions of empty areas on the logical screen as strut values use absolute coordinates, but to set the size of `bar.Gap` correctly we need coordinates relative to physical screen edges. Basically those values are the sole reason for this PR :). Maybe there is a more suggestive name for this?

Reply:
Sorry if I'm being a bit slow here :p  I had to read it a few times to understand what it's doing but I think I get it. So you're starting off with just the x and y values of the edges and then adjusting them using `strut`. I wonder if we can write it so that is more obvious just from a glance, just to avoid confusion in the future. Was it different before? I don't remember getting tripped up on something like this before lol

Reply:
So what I came up with is this. If I understood it correct, then I think we don't have to handle a strut value being 0 any differently. This means we can just do this:
```suggestion
            self.reserved_space = [
                strut[0] - self.screen.x,
                strut[1] - (x_screen_dimensions.width - self.screen.x - self.screen.width),
                strut[2] - self.screen.y,
                strut[3] - (x_screen_dimensions.height - self.screen.y - self.screen.height),
            ]
```
The reason for this is that `strut[0]` will be `0` when a window wants its left edge to be at x == 0 in X screen coods, which I guess could still happen if that window "belongs" to a physical screen whose X coord is not 0. This would be like reserving negative space (if it was attached to that other output... in reality the X server doesn't really bind windows to a specific output). Another example would be if a window that is on an output with its X coordinate at 1920 set `strut[0]` to 1920, then it wants to reserve no extra space i.e. it would be up against that output's left physical edge. In that case, if `strut[0]` actually equalled 0 then it actually wants to be on a different screen than it is.

Anyway I hope my description make sense and I hope that I have understood things correctly! Please correct me if I have misunderstood how the data is encoded here.

Reply:
> strut[0] will be 0 when a window wants its left edge to be at x == 0 in X screen coods

I'm afraid this interpretation is at odds with the ewmh spec. `strut[0:4]` contains
> 4 cardinals specifying the width of the reserved area at each border of the screen

A value of `0` would thus just mean "no reservation for this cardinal". 
In the examples only one cardinal gets specified, which would be insufficient otherwise. Also I don't think the ending remark 

> A "corner" panel that does not extend for the full length of a screen border SHOULD only set one strut.

would not make sense if an unset strut had a meaning apart from "nothing to do here".

Reply:
> Was it different before?

During this PR the list has only changed it's name, but as for the latest release the pure strut values are passed on to be reserved possibly resulting in a too large `bar.Gap`

Reply:
> A value of 0 would thus just mean "no reservation for this cardinal".

so a value of 0 for the left hand side would say "put my left edge against the far left end of the x screen", right? Or is this not the same as "don't reserve me space" (I guess this only holds true if the remaining "partial" cardinals aren't factored in i.e. the bar reserves the entire edge).

Reply:
^ This would be for a horizontal bar on the top or bottom edge

Reply:
Ah, wait. Are we talking about the first 4 cardinals or the last 8? I've been referring to the first 4 the whole time. 

> if the remaining "partial" cardinals aren't factored in i.e. the bar reserves the entire edge

Those partial cardinals are only considered when looking for the correct physical screen. For all I know there is no `bar.Gap` covering only parts of an edge, right?

Reply:
Yeah sorry that's what I meant -- I also meant the first four, just in that hypothetical where a window reserves a band of space on one edge (a band that covers the entire length of that edge), the ends of the reserved area will touch the two adjecent perpendicular edges.

Let's say we have three outputs side-by-side (1, 2 & 3; each are 1000 pixels wide) and we open a window on the middle output and it wants to reserve 20 pixels at the top of that middle output, to look like this:
```
+-------+-------+-------+
|       +-------+       |
|       |       |       |
|       |       |       |
|   1   |   2   |   3   |
+-------+-------+-------+

```
In output-relative coordinates, the window wants to reserve `[0, 0, 20, 0]` (L, R, T, B). What would the strut values provided by the window be?

Reply:
In your example 1, 2 & 3 all have the same height either, right? That would then result in `[0, 0, 20, 0, 0, 0, 0, 0, 1000, 2000, 0, 0]`.
But suppose that 1 & 3 both are 2000px tall, screen 2 just 1800. The three of them are aligned on their bottom edges. Here we'd get [0, 0, 220, 0, 0, 0, 0, 0, 1000, 2000, 0, 0] for a third party bar with the size of 20 pixels.

Reply:
Ahhhh OK, I see, gotcha. Great, thanks for bearing with me!

In that case it all looks good. Was there anything else left to do on the PR?

Reply:
Glad I could help. No, I don't think I had anything else in mind.